packages/alfa-style/src/property/background.ts

```diff
@@ -70,45 +126,712 @@ export namespace Background {
    */
-  export const Image: Property<Image.Specified, Image.Computed> = Property.of(
-    List.of([Keyword.of("none")], ", "),
-    map(
-      separatedList(
-        either(Keyword.parse("none"), css.Image.parse),
-        delimited(option(Token.parseWhitespace), Token.parseComma)
+
+  const parseAttachment = Keyword.parse("fixed", "local", "scroll");
+
+  const parseAttachmentList = map(
+    separatedList(
+      parseAttachment,
+      delimited(option(Token.parseWhitespace), Token.parseComma)
+    ),
+    (repeats) => List.of(repeats, ", ")
+  );
+
+  export const Attachment: Property<
+    Attachment.Specified,
+    Attachment.Computed
+  > = Property.of(
+    List.of([Keyword.of("scroll")], ", "),
+    parseAttachmentList,
+    (style) => style.specified("background-attachment")
+  );
+
+  export type Attachment = Attachment.Specified | Attachment.Computed;
+
+  export namespace Attachment {
+    export type Specified = List<Specified.Item>;
+
+    export namespace Specified {
+      export type Item = Keyword<"fixed" | "local" | "scroll">;
+    }
+
+    export type Computed = Specified;
+  }
+
+  /**
+   * @see https://drafts.csswg.org/css-backgrounds/#propdef-background-position
+   */
+
+  const parsePositionComponent = either(
+    css.Position.parseCenter,
+    either(Length.parse, Percentage.parse)
+  );
+
+  const parsePositionComponentList = map(
+    separatedList(
+      parsePositionComponent,
+      delimited(option(Token.parseWhitespace), Token.parseComma)
+    ),
+    (positions) => List.of(positions, ", ")
+  );
+
+  const parsePosition: Parser<
+    Slice<Token>,
+    [Position.X.Specified.Item, Position.Y.Specified.Item],
+    string
+  > = map(
+    any<
+      Slice<Token>,
+      ["x", Position.X.Specified.Item] | ["y", Position.Y.Specified.Item],
+      string
+    >(
+      map(
+        delimited(
+          option(Token.parseWhitespace),
+          either(
+            map(Keyword.parse("left", "right", "center"), (x) =>
+              x.value === "center" ? x : css.Position.Side.of(x)
+            ),
+            either(Length.parse, Percentage.parse)
+          )
+        ),
+        (x) => ["x", x]
       ),
-      (images) => List.of(images, ", ")
+      map(
+        delimited(
+          option(Token.parseWhitespace),
+          either(
+            map(Keyword.parse("top", "bottom", "center"), (x) =>
+              x.value === "center" ? x : css.Position.Side.of(x)
+            ),
+            either(Length.parse, Percentage.parse)
+          )
+        ),
+        (y) => ["y", y]
+      )
+    ),
+    (positions) => {
+      let x: Position.X.Specified.Item = Keyword.of("center");
+      let y: Position.Y.Specified.Item = Keyword.of("center");
+
+      for (const position of positions) {
+        switch (position[0]) {
+          case "x":
+            x = position[1];
+            break;
+          case "y":
+            y = position[1];
+        }
+      }
+
+      return [x, y];
```

Comment:
I feel this could be streamlined in (roughly)
```javascript
const result = {x: "center"; y: "center"};
for (const [key, value] of positions) {
  result[key] = value;
}

return [result.x, result.y];
```
Not sure it is that much more readable in that specific case, but it does emphasise the similarity in treatment and the fact that `positions` is an array of `[key, value]`.

Reply:
That however messes up the type inference as `key` will be inferred as `"x" | "y"` and `value` will be inferred as `X.Specified.Item | Y.Specified.Item`. We need `"x"` to be assigned a value of `X.Specified.Item` and `"y"` a value of `Y.Specified.Item` which is only possible if we let TypeScript narrow the tuple type before we destruct it, as currently done.